jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java

```diff
@@ -2227,6 +2209,92 @@ public void setNeedClientAuth(boolean needClientAuth)
+         * at the HTTP level for incorrect SNI (see SecureRequestCustomizer).
+         * @return true if no SNI match is handled as no certificate match, false if no SNI match is handled by
+         *         delegation to the non SNI matching methods.
+         */
+        @ManagedAttribute("Whether the TLS handshake is rejected if there is no SNI host match")
+        public boolean isSniRequired()
+        {
+            return _sniRequired;
+        }
+
+        /**
+         * Set if the default {@link #sniSelect(String, Principal[], SSLSession, String, Collection)} implementation
+         * require an SNI match? Note that if a non SNI handshake is accepted, requests may still be rejected
+         * at the HTTP level for incorrect SNI (see SecureRequestCustomizer).
+         * This setting may have no effect if {@link #sniSelect(String, Principal[], SSLSession, String, Collection)} is
+         * overridden or a non null function is passed to {@link #setSNISelector(SniX509ExtendedKeyManager.SniSelector)}.
+         * @param sniRequired true if no SNI match is handled as no certificate match, false if no SNI match is handled by
+         *                    delegation to the non SNI matching methods.
+         */
+        public void setSniRequired(boolean sniRequired)
+        {
+            _sniRequired = sniRequired;
+        }
+
+        @Override
+        protected KeyManager[] getKeyManagers(KeyStore keyStore) throws Exception
+        {
+            KeyManager[] managers = super.getKeyManagers(keyStore);
+            if (isSniRequired())
+            {
+                if (managers == null || Arrays.stream(managers).noneMatch(SniX509ExtendedKeyManager.class::isInstance))
+                    throw new IllegalStateException("No SNI Key managers when SNI is required");
+            }
+            return managers;
+        }
+
+        /**
+         * @return the custom function to select certificates based on SNI information
+         */
+        public SniX509ExtendedKeyManager.SniSelector getSNISelector()
+        {
+            return _sniSelector;
+        }
+
+        /**
+         * <p>Sets a custom function to select certificates based on SNI information.</p>
+         *
+         * @param sniSelector the selection function
+         */
+        public void setSNISelector(SniX509ExtendedKeyManager.SniSelector sniSelector)
+        {
+            _sniSelector = sniSelector;
+        }
+
+        @Override
+        public String sniSelect(String keyType, Principal[] issuers, SSLSession session, String sniHost, Collection<X509> certificates) throws SSLHandshakeException
+        {
+            if (sniHost == null)
+            {
+                // No SNI, find the default certificate to send, or let the JDK decide.
+                return certificates.stream()
+                    .filter(x509 -> Objects.equals(x509.getAlias(), getCertAlias()))
```

Comment:
I think this is not needed now, since we wrap `sniKM(aliasKM(jdkKM))` and `aliasKM` will pick the `certAlias` certificate correctly, so it should be enough to return `DELEGATE`.

Reply:
```java
if (sniHost) {
  return _sniRequired ? null : DELEGATE;
}
```